File: code/_ROS_Generator/Printer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ROS_Generator
{
    class Printer
    {

        // Iteratively print information on all ros packages
        public void print_ros_packages(List<ROS_Package> ros_packages)
        {
            foreach (var package in ros_packages)
            {            
                Console.Out.WriteLine("\nROS Package: " + package.name);
                Console.Out.WriteLine("-----------");

                // Print all msg files in package
                if (package.messages.Count > 0) 
                {
                    Console.Out.WriteLine("\nMessages:");
                    Console.Out.WriteLine("--------");
                    foreach (var msg in package.messages)
                    {
                        Console.Out.WriteLine("* " + msg.name);
                    }
                }

                // Print all msg files in package
                if (package.services.Count > 0)
                {
                    Console.Out.WriteLine("\nServices:");
                    Console.Out.WriteLine("--------");
                    foreach (var srv in package.services)
                    {
                        Console.Out.WriteLine("* " + srv.name);
                    }
                }
              
                // Print all nodes in package
                if (package.nodes.Count > 0)
                {
                    Console.Out.WriteLine("\nNodes: ");
                    Console.Out.WriteLine("-----");

                    foreach (var node in package.nodes)
                    {
                        Console.Out.WriteLine("* " + node.name);
                        print_ros_node(node.components);
                    }
                }           
            }

        }


        // Iteratively print information on all components in ros node
        public void print_ros_node(List<Visitor.ROS_Component> ros_components)
        {
            Console.Out.WriteLine("\n");
            foreach (var component in ros_components)
            {
                // Print Component Name
                Console.Out.WriteLine("Component: " + component.name);
                Console.Out.WriteLine("---------");

                // Print Provided Services
                if (component.provided_services.Count != 0)
                {
                    Console.Out.WriteLine("Provided Services: ");
                    Console.Out.WriteLine("-----------------");
                    foreach (var service in component.provided_services)
                    {
                        if (service.type == "Provides")
                        {
                            Console.Out.WriteLine("* " + service.name);
                        }
                    }
                }

                // Print Required Services
                if (component.required_services.Count != 0)
                {
                    Console.Out.WriteLine("Required Services: ");
                    Console.Out.WriteLine("-----------------");
                    foreach (var service in component.required_services)
                    {
                        if (service.type == "Requires")
                        {
                            Console.Out.WriteLine("* " + service.name);
                        }
                    }
                }

                // Print Publishers
                if (component.publishers.Count != 0)
                {
                    Console.Out.WriteLine("Publishers: ");
                    Console.Out.WriteLine("----------");
                    foreach (var publishers in component.publishers)
                    {
                        Console.Out.WriteLine("* Port: " + publishers.name + " on Topic: " + publishers.topic);
                    }
                }

                // Print Subscribers
                if (component.subscribers.Count != 0)
                {
                    Console.Out.WriteLine("Subscribers: ");
                    Console.Out.WriteLine("-----------");
                    foreach (var subscriber in component.subscribers)
                    {
                        Console.Out.WriteLine("* Port: " + subscriber.name + " on Topic: " + subscriber.topic);
                    }
                }

                // Print Timers
                if (component.timers.Count != 0)
                {
                    Console.Out.WriteLine("Timers: ");
                    Console.Out.WriteLine("------");
                    foreach (var timer in component.timers)
                    {
                        Console.Out.WriteLine("* Timer: " + timer.name +
                                              "; Period: " + timer.period +
                                              " " + timer.period_unit +
                                              "; Offset: " + timer.offset +
                                              " " + timer.offset_unit);
                    }
                }
                Console.Out.WriteLine("\n");
            }
        }
    }
}

File: code/ros_generator/python/printer.py
# Printer
#
# Print information about a ROS workspace
# in: listener object
#
# Author: Pranav Srinivas Kumar
# Date: 2014.12.16

class Printer:

    # Print information about the ROS workspace
    def print_ros_workspace(self, listener):

	print "Workspace: ", listener.workspace.name
	print "---------\n"
	# For each package in the workspace
	for package in listener.workspace.packages:
	    # Print package name
	    print "Package: ", package.name
	    print "-------\n"

	    # For each message in the package
	    if package.messages != []:
	        for message in package.messages:
		    print "Message: ", message.name
		    print "-------"
                    # Print all message fields
		    for field in message.fields:
		        print "* ", field[0], field[1]
		    print "\n"

	    # For each service in the package
	    if package.services != []:
	        for service in package.services:
		    print "Service: ", service.name
		    print "-------"
                    # Print all service request fields
		    if service.request_fields != []:
		        print "Request Fields:"
		        print "--------------"
		        for req in service.request_fields:
			    if len(req) > 2:
			        print "* ", req[0], req[1], " = ", req[2]
			    else:
			        print "* ", req[0], req[1]
                    # Print all service response fields
		    if service.response_fields != []:
		        print "Response Fields:"
		        print "---------------"
		        for res in service.response_fields:
			    if len(res) > 2:
			        print "* ", res[0], res[1], " = ", res[2]
			    else:
			        print "* ", res[0], res[1]
		    print "\n"
            
            # For each component in the package
            if package.components != []:
                for component in package.components:
                    print "Component: ", component.name
                    print "---------"
                    # Print all provided services
                    if component.provided_services != []:
                        print "Provided Services: "
                        print "-----------------"
                        for provided in component.provided_services:
                            print "* ", provided
                    # Print all required services
                    if component.required_services != []:
                        print "Required Services: "
                        print "-----------------"
                        for required in component.required_services:
                            print "* ", required
                    # Print all publisher ports
                    if component.publishers != []:
                        print "Publishers: "
                        print "----------"
                        for publisher in component.publishers:
                            print "* Name: ", publisher.name, "; Topic: ", publisher.topic
                    # Print all subscriber ports
                    if component.subscribers != []:
                        print "Subscribers: "
                        print "-----------"
                        for subscriber in component.subscribers:
                            print "* Name: ", subscriber.name, "; Topic: ", subscriber.topic
                    # Print all component timers
                    if component.timers != []:
                        print "Timers: "
                        print "------"
                        for timer in component.timers:
                            print "* ", timer.name, "; Period: ", timer.period, timer.period_unit
                    print "\n"
            
                    
            # For each node in the package
            if package.nodes != []:
                for node in package.nodes:
                    print "Node: ", node.name
                    print "----"
                    for component in node.components:
                        print "* Instance: ", component[1], "; Type: ", component[0]
                    print "\n"
